src/historywork/BatchDownloadWork.cpp

```diff
@@ -14,133 +14,58 @@
 
 namespace stellar
 {
-
-BatchDownloadWork::BatchDownloadWork(Application& app, WorkParent& parent,
-                                     CheckpointRange range,
+BatchDownloadWork::BatchDownloadWork(Application& app, CheckpointRange range,
                                      std::string const& type,
                                      TmpDir const& downloadDir)
-    : Work(app, parent,
-           fmt::format("batch-download-{:s}-{:08x}-{:08x}", type, range.first(),
-                       range.last()))
+    : BatchWork(app, fmt::format("batch-download-{:s}-{:08x}-{:08x}", type,
+                                 range.first(), range.last()))
     , mRange(range)
-    , mNext(mRange.first())
+    , mNext(range.first())
     , mFileType(type)
     , mDownloadDir(downloadDir)
-    , mDownloadCached(app.getMetrics().NewMeter(
-          {"history", "download-" + type, "cached"}, "event"))
-    , mDownloadStart(app.getMetrics().NewMeter(
-          {"history", "download-" + type, "start"}, "event"))
-    , mDownloadSuccess(app.getMetrics().NewMeter(
-          {"history", "download-" + type, "success"}, "event"))
-    , mDownloadFailure(app.getMetrics().NewMeter(
-          {"history", "download-" + type, "failure"}, "event"))
-{
-}
-
-BatchDownloadWork::~BatchDownloadWork()
 {
-    clearChildren();
 }
 
 std::string
 BatchDownloadWork::getStatus() const
 {
-    if (mState == WORK_RUNNING || mState == WORK_PENDING)
+    if (getState() == State::WORK_RUNNING)
     {
         auto task = fmt::format("downloading {:s} files", mFileType);
         return fmtProgress(mApp, task, mRange.first(), mRange.last(), mNext);
     }
-    return Work::getStatus();
+    return BatchWork::getStatus();
 }
 
-void
-BatchDownloadWork::addNextDownloadWorker()
+std::shared_ptr<BasicWork>
+BatchDownloadWork::yieldMoreWork()
 {
-    if (mNext > mRange.last())
+    if (!hasNext())
     {
-        return;
+        CLOG(WARNING, "Work")
+            << getName() << " has no more children to iterate over! ";
+        return nullptr;
     }
 
     FileTransferInfo ft(mDownloadDir, mFileType, mNext);
-    if (fs::exists(ft.localPath_nogz()))
-    {
-        CLOG(DEBUG, "History")
-            << "already have " << mFileType << " for checkpoint " << mNext;
-        mDownloadCached.Mark();
-    }
-    else
-    {
-        CLOG(DEBUG, "History") << "Downloading and unzipping " << mFileType
-                               << " for checkpoint " << mNext;
-        auto getAndUnzip = addWork<GetAndUnzipRemoteFileWork>(ft);
-        assert(mRunning.find(getAndUnzip->getUniqueName()) == mRunning.end());
-        mRunning.insert(std::make_pair(getAndUnzip->getUniqueName(), mNext));
-        mDownloadStart.Mark();
-    }
+    CLOG(DEBUG, "History") << "Downloading and unzipping " << mFileType
+                           << " for checkpoint " << mNext;
+    auto getAndUnzip = addWork<GetAndUnzipRemoteFileWork>(ft);
```

Comment:
I think it is strange to make implementors of yieldMoreWork do the addWork themselves.

I think that:
1. we should replace that addWork with non-templated addWork that just gets shared_ptr<BasicWork> (you got rid of passing *this to new work instances, so this wont hurt)
2. caller of yieldMoreWork should then do addWork() on result (not it justs adds result to mBatch).
3. note: interesting, Work::addChild does not check for duplicate work, but BatchWork checks for that